### DISCUSSION TITLE
Symfony Form : CollectionType, "allow_add: true", and access to "$options['data']" in the subform

### DISCUSSION THREAD

**Discussion by Rom1906** [symfony/symfony#58481] [Q&A]
Hello I am developing a web application using Symfony 7, and I am encountering a specific issue when working with nested forms and the CollectionType field.

### Use Case

- I have a form A that contains a "CollectionType" field called "businesses_allocation."
- This field represents a collection of subforms B. Each subform B needs to receive a "Business" entity as an option in order to manipulate this entity and dynamically create the fields of the form based on a collection present in the "Business" entity.
- By default, the collection should be empty when the page loads. Then, a button allows users to dynamically add items to this collection via an AJAX request.
- I want to use the "allow_add: true" parameter on the "CollectionType" field to allow the dynamic addition of new subforms.

### Issue

I want to access the data passed in the "$options['data']" option in each subform of the collection. However, when I try to do so via the "PRE_SET_DATA" event in the subform, I receive the following error:

`Warning: Undefined array key "data"`

This happens when new items are added to the collection, as the data is not always available or defined at this stage.

**Comment by siganushka** [comment]
`$options['data']` will receive the data passed when creating the form, e.g:

```php
$entity = new YourEntity
$form = $this->createForm(FooType::class, $entity);

// in your FooType::class
$options['data'] = $entity
```

If no data is passed when creating a form, the `$options['data']`  is `NULL`.

If you want to get the current data inside the form (including new data that has been entered), you need to use the form event:

```php
// in your FooType::class

public function buildForm(FormBuilderInterface $builder, array $options): void
{
    $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) {
        // your current new data
        $newData = $event->getData();
        // data passed when creating a form
        $originData = $event->getForm()->getData();
    })
}
```

In your example you need to use the event in `businesses_allocation`.
